lib/attrsets.nix

```
@@ -1208,6 +1208,102 @@ rec {
     in
     recurse [ ] set;
 
+  /**
+    Recursively apply a function to each leaf attribute of the potentially
+    nested attribute set and combine the results in a list.
+
+    The first argument to the mapping function is a list of the attribute names
+    that form the path to the leaf attribute. The second argument will never
+    be an attribute set.
+
+    # Inputs
+
+    `f`
+
+    : A function, given an attribute’s path and value, returns a new value.
+
+    `attrs`
+
+    : Attribute set to map over.
+
+    # Type
+
+    ```
+    mapAttrsToListRecursive :: ([String] -> a -> b) -> AttrSet -> [b]
+    ```
+
+    # Examples
+    :::{.example}
+    ## `lib.attrsets.mapAttrsToListRecursive` usage example
+
+    ```nix
+    mapAttrsToListRecursive (path: value: "${concatStringsSep "." path}=${value}")
+      { n = { a = "A"; m = { b = "B"; c = "C"; }; }; d = "D"; }
+    => [ "n.a=A" "n.m.b=B" "n.m.c=C" "d=D" ]
+    ```
+    :::
+  */
+  mapAttrsToListRecursive =
+    f:
+    set:
+    mapAttrsToListRecursiveCond (path: as: true) f set;
+
+  /**
+    Like `mapAttrsToListRecursive`, but it takes an additional predicate to
+    decide whether to recurse into an attribute set. If the predicate returns
+    false, `mapAttrsToListRecursiveCond` does not recurse, but instead applies
+    the mapping function. If the predicate returns true, it does recurse
+    without applying the mapping function.
+
+    The first argument to the predicate is a list of the attribute names
+    that form the path to the leaf attribute. The second argument is the
+    attribute set.
+
+    # Inputs
+
+    `pred`
+
+    : Predicate to decide whether to recurse into an attribute set.
+
+    `f`
+
+    : A function, given an attribute’s path and value, returns a new value.
+
+    `attrs`
+
+    : Attribute set to map over.
+
+    # Type
+    ```
+    mapAttrsToListRecursiveCond :: ([String] -> AttrSet -> Bool) -> ([String] -> a -> b) -> AttrSet -> [b]
```

```suggestion
    mapAttrsToListRecursiveCond :: ([String] -> AttrSet -> Bool) -> ([String] -> a -> b) -> a -> [b]
```

---

I am not sure if I understand the notation correctly, but isn’t the argument always an attribute set?

---

At least in your tests it is. You could add `assert isAttrs`. At least what happens if we pass something else?

Roberts suggestion should really be:

```suggestion
    mapAttrsToListRecursiveCond :: ([String] -> a -> Bool) -> ([String] -> a -> b) -> a -> [b]
```

Using `a` as a type-variable. So you see that both the first and the second function receive the `a` that was passed as third argument. Its only a convention and those rules are not written down yet.